hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/DynamoDBBasedLockProvider.java

```diff
@@ -0,0 +1,226 @@
+@NotThreadSafe
+public class DynamoDBBasedLockProvider implements LockProvider<LockItem> {
+
+  private static final Logger LOG = LogManager.getLogger(DynamoDBBasedLockProvider.class);
+
+  private final AmazonDynamoDBLockClient client;
+  private final long leaseDuration;
+  private final String tableName;
+  protected LockConfiguration lockConfiguration;
+  private volatile LockItem lock;
+
+  public DynamoDBBasedLockProvider(final LockConfiguration lockConfiguration, final Configuration conf) {
+    checkRequiredProps(lockConfiguration);
+    this.lockConfiguration = lockConfiguration;
+    this.tableName = lockConfiguration.getConfig().getString(DYNAMODB_TABLE_NAME_PROP_KEY);
+    this.leaseDuration = Long.parseLong(lockConfiguration.getConfig().getString(LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY));
+    AmazonDynamoDB dynamoDB = getDynamoClient();
+    // build the dynamoDb lock client
+    this.client = new AmazonDynamoDBLockClient(
+        AmazonDynamoDBLockClientOptions.builder(dynamoDB, tableName)
+                .withTimeUnit(TimeUnit.MILLISECONDS)
+                .withLeaseDuration(leaseDuration)
+                .withHeartbeatPeriod(leaseDuration / 3)
+                .withCreateHeartbeatBackgroundThread(true)
+                .build());
+    if (!client.lockTableExists()) {
+      createLockTableInDynamoDB(
+              dynamoDB,
+              tableName,
+              lockConfiguration.getConfig().getString(DYNAMODB_BILLING_MODE_PROP_KEY),
+              Long.parseLong(lockConfiguration.getConfig().getString(DYNAMODB_READ_CAPACITY_PROP_KEY)),
+              Long.parseLong(lockConfiguration.getConfig().getString(DYNAMODB_WRITE_CAPACITY_PROP_KEY)));
+    }
+  }
+
+  // Only used for testing
+  public DynamoDBBasedLockProvider(final LockConfiguration lockConfiguration, final AmazonDynamoDBLockClient client) {
+    checkRequiredProps(lockConfiguration);
+    this.lockConfiguration = lockConfiguration;
+    this.tableName = lockConfiguration.getConfig().getString(DYNAMODB_TABLE_NAME_PROP_KEY);
+    this.leaseDuration = Long.parseLong(lockConfiguration.getConfig().getString(LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY));
+    this.client = client;
+  }
+
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    LOG.info(generateLogStatement(LockState.ACQUIRING, generateLogSuffixString()));
+    long millisTime = unit.toMillis(time);
+    try {
+      lock = client.acquireLock(AcquireLockOptions.builder(lockConfiguration.getConfig().getString(DYNAMODB_PARTITION_KEY_PROP_KEY))
+              .withAdditionalTimeToWaitForLock(millisTime - leaseDuration > 0 ? millisTime - leaseDuration : 0)
+              .withTimeUnit(TimeUnit.MILLISECONDS)
+              .build());
+      LOG.info(generateLogStatement(LockState.ACQUIRED, generateLogSuffixString()));
+    } catch (InterruptedException | LockNotGrantedException e) {
+      throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_ACQUIRE, generateLogSuffixString()), e);
```

Comment:
> If lock is not granted, why do we throw exception here? LockManager has a retry mechanism through which it will try to acquire lock few times with some delays inbetween. Don't think we can throw from here if lock could not be acquired.

If we don't catch ```LockNotGrantedException```, then ```LockManager``` also won't catch this exception: https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/LockManager.java#L75, then the whole program would crush w/o retry logic.

Reply:
If we don't catch ```LockNotGrantedException```, then ```LockManager``` also won't catch this exception: https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/LockManager.java#L75, then the whole program would crush w/o retry logic.

Reply:
what I meant is, we should catch LockNotGrantedException and return false. so that caller might retry again.

Reply:
I see your point. changed.